onc-healthit/lantern-back-end
# Org ids names sources 1

**Included Changes**
* Adds npi_ids onto the fhir_endpoints db table and data structure
* npi ids are added to the npi_ids list if an endpoint points to multiple npi ids (from the npi contacts loader)
* make matching algorithm pay attention to npi_ids (if they exist)
* makes organization_name be organization_names in db table and data structure
  * any repeated entries has the organization name added to the organization_names list
* gives match priority to matches from the npi_ids by reducing the confidence of all name-based matches by 10%

**Note:** the `populatedb` script has been changed so the NPI contacts list is loaded before the NPI organization list, so if you're on auto-pilot, you'll need to be careful of that change.

**Testing**

The database needs to be cleaned before running because there are changes to the database schemas.

```
make clean
make test_all
```

**Manual Testing**

We don't care about the capability statements, so we don't need to worry about running the capability sender/receiver code, just the basic loading code.

Populate the DB with a 100k test file.

```
make populatedb
```

*Check multiple organizations*

Connect to the db:

```
docker exec -it lantern-back-end_pg_prometheus_1 psql -U lantern
```

Browse through the `fhir_endpoints` table to confirm that there are entries where multiple organizations are stored. Example is "https://emrproxy.mcfarlandclinic.com/FHIRProxy/api/FHIR/DSTU2/":

```
SELECT url, organization_names FROM fhir_endpoints;
```

*Check multiple IDs*

Look at the `fhir_endpoints` entries from NPPES and note that the NPI_IDs are populated and that some have multiple NPI_IDs.

```
SELECT url, npi_ids, organization_names FROM fhir_endpoints WHERE list_source='NPPES';
```

*Check linking against names*

Browse the endpoint/organization matches that are not from NPPES and confirm that they look sane and that none have a "1.0" confidence:

```
SELECT url, list_source, endpoint_names, organization_name,  match_score FROM org_mapping;
```

*Check linking against IDs*

Unfortunately, the NPI IDs that we have in our endpoints list seem to be for individuals rather than for organizations, so there is not a link within our database. To test that the algorithm is working in addition to running the unit tests, do the following:

```
INSERT INTO fhir_endpoints (url, npi_ids, list_source) VALUES ('https://www.lantern.example.com', '{"1023011079"}', 'NPPES');
```

Then in the terminal, run the linker

```
cd endpointmanager/cmd/endpointlinker
go run main.go
```

Back in the DB view, run this query:

```
SELECT url, list_source, endpoint_names, organization_name, match_score FROM org_mapping WHERE list_source='NPPES';
```

You should see:

![image](https://user-images.githubusercontent.com/11574199/82445315-941b7c80-9a72-11ea-88c4-9e30327d3167.png)


The additional entries are because a that URL is part of other lists which include an organization name that had a match.

-----

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-152
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.
